new docs for V5 (#186)
buildo/avenger: README.md

## queries
The [**`query`**](#query) function allows you to query your data source and get an object of type [**`CachedQuery`**](#CachedQuery) in return.
It accepts two parameters, the first is a function with a [**`Fetch`**](#Fetch) signature that is used to retrive data from your data source, the second is an object with the[**`Strategy`**](#Strategy) signature that will be used to decide if the data stored by **Avenger** is still relevant or needs to be refetched.

Although important, `query` is a pretty low level API and **Avenger** offers some convenient utils with a [**`StrategyBuilder`**](#StrategyBuilder) signature that you should prefer over it (unless you have very specific needs):

- **refetch:** runs the fetch function everytime the data is requested.
- **expire:** when the data is requested, the fetch function is run only if data in the `Cache` is older than the expiration defined, otherwise the cached value is used.
- **available:** when the data is requested, if a cahed value is available it is always returned, otherwise the fetch function is run and the result stored in the `Cache` accordingly.

> All these utils ask you to specify custom `Setoid` instances for the inputs and outputs of the `Fetch` function but you can (and should) use them toghether with one of the built in implementations:

Reviewer: Not clear the `why` and it requires to know what a `Setoid` is.
"will use *** equality check" -> to check what?

Author: I put a link to `fp-ts` docs for Setoid and tried to be a little clearer on what they are used for

- **queryShallow:** will use a shallow equality check.
- **queryStrict:** will use a strict equality check.

> - **queryJSON:** will use a strict equality check after transforming the data via JSON stringification.

Author: do you think they are still needed after the rewrite?

Some examples will help clarify:
```ts
/*
  this implementation will always re-run the `Fetch` function

>   even is valid cached data is already present

Reviewer: ```suggestion
  even if valid cached data is already present
```

  and use shallow equality to compare input output when called
*/
const myQuery = queryShallow(fetchFunction, refetch);

/*
  this implementation will never run the `Fetch` function
  unless no valid data is present in the Cache

>   and use strict equality to compare input output when called

Reviewer: ```suggestion
  and use strict equality to compare input/output when called
```

*/
const myQuery = queryStrict(fetchFunction, available);

/*
  this implementation will run the `Fetch` function only if no valid data is present in the Cache
  or t > 10000 ms passed till the last time data was fetched
  and use strict equality to compare input output when called
*/
const myQuery = queryJSON(fetchFunction, expire(10000));
```

> Each time the `Fetch` function is run with some `inputs`, those same `inputs` are used as `key` to store the result obtained; from that moment onwards when **Avenger** will need to decide if the data in our [**`Cache`**](#Cache) is present and still valid it will first attempt to retrive it data from the [**`Cache`**](#Cache) by performing an equality check between the `inputs` used to call the `Fetch` function and the keys of the `CacheMap` and then match the result against the cache strategy defined (for instance if we chose `refetch` the data will always be deemed invalid irrespective of the result).

Reviewer: ```suggestion
Each time the `Fetch` function is run with some `inputs`, those same `inputs` are used as `key` to store the result obtained; from that moment onwards when **Avenger** will need to decide if the data in our [**`Cache`**](#Cache) is present and still valid it will first attempt to retrieve its data from the [**`Cache`**](#Cache) by performing an equality check between the `inputs` used to call the `Fetch` function and the keys of the `CacheMap` and then match the result against the cache strategy defined (for instance if we chose `refetch` the data will always be deemed invalid irrespective of the result).
```

> Each time the `Fetch` function is run with some `inputs`, those same `inputs` are used as `key` to store the result obtained; from that moment onwards when **Avenger** will need to decide if the data in our [**`Cache`**](#Cache) is present and still valid it will first attempt to retrive it data from the [**`Cache`**](#Cache) by performing an equality check between the `inputs` used to call the `Fetch` function and the keys of the `CacheMap` and then match the result against the cache strategy defined (for instance if we chose `refetch` the data will always be deemed invalid irrespective of the result).

Reviewer: 🤯rephrase?

If the result is valid it is returned, otherwise the `Fetch` function will be re-run in order to get new and valid data.

> There are two ways get a query result:

Reviewer: ```suggestion
There are two ways to get a query result:
```

```ts

> type Error = 'there was an error';

Reviewer: this is a strange type for an `Error`

type User = { userName: String };

> declare getUser(userId: number): TaskEither<Error, User>

Reviewer: ```suggestion
declare function getUser(userId: number): TaskEither<Error, User>
```

> const userQuery: ObservableQuery<number, Error, User> = query(

Reviewer: which `Setoid` is used here calling `query` directly?

  getUser,
  refetch
);

declare function dispatchError(e: Error): void;
declare function setCurrentUser(e: User): void;

// feeding your query to `observe` will give you an observable on the query
// N.B. until now no fetch is yet attempted, avenger will wait until the first subscription is issued
const observable: Observable<QueryResult<Error, User>> = observe(userQuery);

> // after this the fetch function triggers

Reviewer: maybe:
```suggestion
// this will trigger the fetch function
```

observable.subscribe(dispatchError, setCurrentUser);

// alternatively you can call `run` on your query and it will return a TaskEither<Error, User>
// you can then use it imperatively
const task: TaskEither<Error, User> = userQuery.run(1);
const result: Either<Error, User> = await task.run();
```

altough the `run` method is available to check a query result imperatively, it is highly suggested the use of the `observe` utility in order to be notified in real time of when data changes. Either way, whenever you ask for a query result you will end up with an object with the [**`QueryResult`**](#QueryResult) signature that conveniently lets you fold to decide the best way to handle the result.

the flow when you subscribe / call run on a cached value is relatively simple:
!["cached flow"](docs/CachedValue.svg)

> when you call run / subscribe on a combination `query`/`inputs` that was never retrieved before (or whose last try to retrive it ended up with a `Failure`), avenger will try to run the `Fetch` function resulting in a more complex flow:

Reviewer: search  & replace pls:
```suggestion
when you call run / subscribe on a combination `query`/`inputs` that was never retrieved before (or whose last try to retrieve it ended up with a `Failure`), avenger will try to run the `Fetch` function resulting in a more complex flow:
```

## composing queries
you can build bigger queries from smaller ones in two ways:

- by composing them with [**`compose`**](#compose): when you need your queries to be sequentially run with the results of one feeding the other, you can use `compose`:

```ts
import { compose } from 'avenger/lib/Query';

type UserPreferences = { color: string };

// note that the two ends of the composed functions must have compatible types
declare function getUser(userId: number): TaskEither<Error, User>;
declare function getUserPreferences(
  user: User
): TaskEither<Error, UserPreferences>;

const userQuery: ObservableQuery<number, Error, User> = queryShallow(
  getUser,
  refetch
);

const preferencesQuery: ObservableQuery<
  User,
  Error,
  UserPreferences
> = queryShallow(getUserPreferences, refetch);

const composition: Composition<number, Error, UserPreferences> = compose(
  userQuery,
  preferencesQuery
);
```

- by grouping them with [**`product`**](#product): when you don't need to run the queries sequentially but would like to conveniently group them and treat them as if they where one you can use `product`:

```ts
const grouped = product({ myQuery, myQuery2 });
```

# commands

> Since now we only described how fetching of data is performed, when you need to update / insert data remotely you can make use of [**`command`**](#command):

Reviewer: ```suggestion
Since now we only described how fetching of data is performed. When you need to update / insert data remotely you can make use of [**`command`**](#command):
```

```ts
declare function updateUserPreferences({
  color: string
}): TaskEither<Error, void>;

const updatePreferencesCommand = command(updateUserPreferences, {
  preferencesQuery
});
```

> `command` accepts `Fetch` function that will be used to modify the remote data source and a second parameter, a `Record` of `query`es that will be invalidated once the command is succesfully run:

Reviewer: ```suggestion
`command` accepts a `Fetch` function that will be used to modify the remote data source and a second parameter, a `Record` of `query`es that will be invalidated once the command is succesfully run:
```

```ts
/* when you call the command you can specify the input value corresponding

> to the Cache key that sould be invalidated as a second parameter */

Reviewer: ```suggestion
to the Cache key that should be invalidated as a second parameter */
```

> updatePreferencesCommand({ color: 'acquamarine' }, { preferencesQuery: 1 });

Reviewer: What's the key if `preferencesQuery` has more than one input parameter?

Author: the `Fetch` function is unary so you can only have 1 parameter

```

# React

Avenger also exports some utilities to use with `React`

## declareQueries

> `declareQueries` in an `HOC` builder, it lets you define the queries that you want to inject into a component and then creates a simple `HOC` to wrap it:

Reviewer: ```suggestion
`declareQueries` is an `HOC` builder. It lets you define the queries that you want to inject into a component and then creates a simple `HOC` to wrap it:
```
?

```ts
import { declareQueries } from 'avenger/lib/react';
import { userPreferences } from './queries';

const queries = declareQueries({ userPreferences });

class MyComponent extends React.PureComponent<Props, State> {
  render() {

>     return this.props.queries.fold(

Reviewer: Since now we never talked about folding on queries and query states (loading/error/ready)

Author: there is a brief mention when I say "(...) that conveniently lets you fold (...)" and a link to the type of `QueryResult` that mentions it is a bifunctor and monad. I thought that for an experienced reader that would be enough, at the same time experieced readers are probably the only ones interested in this.. I will expand a little on that

      () => <p>loading</p>,
      () => <p>there was a problem when fetching preferences</p>,
      ({ userPreferences }) => <p>my favourite color is {userPreferences.color}</p>
    )
  }
}

export queries(MyComponent)
```

> When using this component from outside you will have to pass it the correct query patameters inside the `queries` prop in order for it to load the declared queries:

Reviewer: ```suggestion
When using this component from outside you will have to pass it the correct query parameters inside the `queries` prop in order for it to load the declared queries:
```

```ts
class MyOtherComponent extends React.PureComponent<Props, State> {
  render() {
    return (
      <MyComponent
        queries={{
          userPreferences: { userName: 'Mario' }
        }}
      />
    );
  }
}
```

## WithQuery

alternatively, to avoid unecessary boilerplate, you can use the `WithQuery` component:

```ts

> import { WithQuery } from 'avenger/lib/react';

Reviewer: missing import of `userPreferences`?

class MyComponent extends React.PureComponent<Props, State> {
  render() {
    return (
      <WithQuery
        query={userPreferences},
        input={{ userName: 'Mario' }}
        render={userPreferences => userPreferences.fold(
          () => <p>loading</p>,
          () => <p>there was a problem when fetching preferences</p>,
          (userPreferences) => <p>my favourite color is {userPreferences.color}</p>
        )}
      />
    )
  }
}
```

# Navigation

> Another useful set of utilities is the one used to handle client navigation in the browser. Following you can find a simple but exhautive example of how it is used:

Reviewer: ```suggestion
Another useful set of utilities is the one used to handle client navigation in the browser. Following you can find a simple but exhaustive example of how it is used:
```

```ts

> export type CurrentView =

Reviewer: Missing imports?

[![](https://travis-ci.org/buildo/avenger.svg)](https://travis-ci.org/buildo/avenger)
[![](https://img.shields.io/npm/v/avenger.svg?sytle=flat-square)](https://www.npmjs.com/package/avenger)
[![npm downloads](https://img.shields.io/npm/dm/avenger.svg?style=flat-square)](https://www.npmjs.com/package/avenger)
[![](https://david-dm.org/buildo/avenger.svg)](https://david-dm.org/buildo/avenger#info=dependencies&view=list)
[![](https://david-dm.org/buildo/avenger/dev-status.svg)](https://david-dm.org/buildo/avenger#info=devDependencies&view=list)

---

# Intro

> Avenger is a data fetching and caching layer written in TypeScript, its API is designed to mirror the principles of **Command Query Responsibility Segregation** thus of facilitating their adoption.

Reviewer 2: ...in TypeScript **dot** Its API is designed to mirror the principles of Command Query Responsibility Segregation ~~thus of facilitating~~  **and to facilitate** their adoption.

If you are new to the concept you can get a grasp of its foundations in [this nice aricle](https://martinfowler.com/bliki/CQRS.html) by Martin Fowler. Using his words:

_"The change that CQRS introduces is to separate models for update and display, which it refers to as "Command" and "Query" respectively following the vocabulary of CommandQuerySeparation._

_The rationale is that for many problems, particularly in more complicated domains, having the same conceptual model for commands and queries leads to a more complex model that does neither well."_

# Avenger
At the heart of the DSL of the software there are two constructors: **query** and **command**.

## queries

> The [**`query`**](#query) function allows you to query your data source and get an object of type [**`CachedQuery`**](#CachedQuery) in return.

Reviewer 3: There's some confusion regarding `CachedQuery` and `ObservableQuery`. I understood only at the end what's going on here.

> It accepts two parameters, the first is a function with a [**`Fetch`**](#Fetch) signature that is used to retrieve data from your data source, the second is an object with the[**`Strategy`**](#Strategy) signature that will be used to decide if the data stored by **Avenger** is still relevant or needs to be refetched.

Reviewer 2: It accepts two parameters **colon** `...`  data source **semicolon** the second `...`

- **queryStrict:** will use a `Setoid` instance that performs a strict equality check to compare inputs.
- **queryJSON:** will use a `Setoid` instance that performs a strict equality check after transforming the data via JSON stringification to compare inputs.

Some examples will help clarify:
```ts
/*
  this implementation will always re-run the `Fetch` function
  even if valid cached data is already present
  and use shallow equality to compare input
*/
const myQuery = queryShallow(fetchFunction, refetch);

/*
  this implementation will never run the `Fetch` function
  unless no valid data is present in the Cache
  and use strict equality to compare input
*/
const myQuery = queryStrict(fetchFunction, available);

/*
  this implementation will run the `Fetch` function only if no valid data is present in the Cache
  or t > 10000 ms passed till the last time data was fetched
  and use strict equality to compare input
*/
const myQuery = queryJSON(fetchFunction, expire(10000));
```

Each time the `Fetch` function is run with some `inputs`, those same `inputs` are used as a `key` to store the result obtained:

> ```

Reviewer 2: missing language `ts`

Author: this is actually pseudo lang not real `ts`

Reviewer 2: 👍

// usersCache is empty
usersCache: {}

//a user is fetched
getUser({ userId: 1 }) -> { userName: "Mario" }

// usersCache is now populated
usersCache: {
  [{ userId: 1 }]: { userName: Mario }
}
```

From that moment onwards, when **Avenger** will need to decide if the data in our [**`Cache`**](#Cache) is present and still valid it will:

1. attempt to retrieve data from the [**`Cache`**](#Cache)
2. match the result against the cache strategy defined (for instance if we chose `refetch` the data will always be deemed invalid irrespective of the result).

If the result is valid it is returned, otherwise the `Fetch` function will be re-run in order to get new and valid data. The two flows are relatively simple:
##### Valid CacheValue

> !["cached flow"](docs/CachedValue.svg)

Reviewer 2: ![image](https://user-images.githubusercontent.com/6418684/60347026-bd47a000-99bc-11e9-93ed-6c70b6c9b49c.png)

Author: I don't get it in the readme it shows correctly.. 😓

Reviewer 2: It could be just an error or the rich diff, ignore it 🤷‍♂

##### Invalid CacheValue
when you call `run` or `subscribe` on a `query` with a combination of `inputs` that was never used before (or whose last use ended up with a `Failure`), avenger will try to run the `Fetch` function resulting in a more complex flow:

> !["cached flow"](docs/UncachedOrErrorValue.svg)

Reviewer 2: Same problem as above with the image

Reviewer 2: Same here, just ignore

There are two ways to get a query result:

```ts
type Error = 'there was an error';
type User = { userName: String };

declare function getUser(userId: number): TaskEither<Error, User>

const userQuery: ObservableQuery<number, Error, User> = query(
  getUser,
  refetch
);

declare function dispatchError(e: Error): void;
declare function setCurrentUser(e: User): void;

// feeding your query to `observe` will give you an observable on the query
// N.B. until now no fetch is yet attempted, avenger will wait until the first subscription is issued
const observable: Observable<QueryResult<Error, User>> = observe(userQuery);

// after this the fetch function triggers

> observable.subscribe(dispatchError, setCurrentUser);

Reviewer 3: What if I want two subscribers? Does the fetch happen twice according to cache semantics?

Author: exactly, it depends on the strategy passed to the `query` util

// alternatively you can call `run` on your query and it will return a TaskEither<Error, User>
// you can then use it imperatively
const task: TaskEither<Error, User> = userQuery.run(1);
const result: Either<Error, User> = await task.run();
```

although the `run` method is available to check a query result imperatively, it is highly suggested the use of the `observe` utility in order to be notified in real time of when data changes.

> Either way, whenever you ask for a query result you will end up with an object with the [**`QueryResult`**](#QueryResult) signature that conveniently lets you fold to decide the best way to handle the result. The `fold` method takes 3 functions as parameters, the first is used to handle a `Loading` result, the second is used in case a `Failure` occurs and the last one handles `Success` values.

Reviewer 2: takes ~~3~~ **three** functions as parameters **colon** the first ... **semicolon** the second ...

> Either way, whenever you ask for a query result you will end up with an object with the [**`QueryResult`**](#QueryResult) signature that conveniently lets you fold to decide the best way to handle the result. The `fold` method takes 3 functions as parameters, the first is used to handle a `Loading` result, the second is used on case a `Failure` occurs and the last one handles `Success` values.

Reviewer 3: It's not clear to me how a subscriber is supposed to listen to the Loading case.

Author: when `observe` is used on a `query` it returns an `observable` with a `.subscribe` method that accepts as an argument a unary function that will be called each time the result changes with the new result. In the docs I kind of assumed that the reader would already know about observables, do you think I should make that part explicit?

## composing queries
You can build bigger queries from smaller ones in two ways:

> - by composing them with [**`compose`**](#compose): when you need your queries to be sequentially run with the results of one feeding the other, you can use `compose`:

Reviewer 2: Pointed list with code example in between is hard to read, you could unify the two sentences in a single one and then say: "here are examples of compose and product respectively:" or something similar.

![image](https://user-images.githubusercontent.com/6418684/60347398-7dcd8380-99bd-11e9-9784-5506a71d0610.png)

```ts
import { compose } from 'avenger/lib/Query';

type UserPreferences = { color: string };

// note that the two ends of the composed functions must have compatible types
declare function getUser(userId: number): TaskEither<Error, User>;
declare function getUserPreferences(
  user: User
): TaskEither<Error, UserPreferences>;

const userQuery: ObservableQuery<number, Error, User> = queryShallow(
  getUser,
  refetch
);

const preferencesQuery: ObservableQuery<
  User,
  Error,
  UserPreferences
> = queryShallow(getUserPreferences, refetch);

const composition: Composition<number, Error, UserPreferences> = compose(
  userQuery,
  preferencesQuery
);
```

- by grouping them with [**`product`**](#product): when you don't need to run the queries sequentially but would like to conveniently group them and treat them as if they where one you can use `product`:

```ts

> const grouped = product({ myQuery, myQuery2 });

Reviewer 3: I would be interesting to see the resulting type. I'd expect some tupled form for example... is that true?

Author: the result is an `ObservableQuery` that is something more "opinionated" than a tupled form, the `queries` of the record are `folded` using a custom logic (is either of  myQuery, myQuery2 is loading then `grouped` is loading). This could actually be interesting to explicit..

```

# commands

> Since now we only described how fetching of data is performed, when you need to update or insert data remotely you can make use of [**`command`**](#command):

Reviewer 2: ~~Since now~~ **Up to now, ** we only described how to fetch data **dot**

```ts
declare function updateUserPreferences({
  color: string
}): TaskEither<Error, void>;

const updatePreferencesCommand = command(updateUserPreferences, {
  preferencesQuery
});
```

> `command` accepts `Fetch` function that will be used to modify the remote data source and a second parameter, a `Record` of `query`es that will be invalidated once the command is succesfully run:

Reviewer 3: So the `Fetch` concept is not strictly related to fetching some data, but to a more general remote call. Is it because of widely used terminology in the frontend?

Author: you are correct, `Fetch` is not a super descriptive name indeed... cc @giogonzo

Reviewer 4: In fp-ts lingo this is just a `ReaderTaskEither`: 3 effects (function/input, async, error) stacked. `fetch` is just the old avenger lingo, introduced a few years ago when describing the API for the first time with Giulio (see https://github.com/buildo/avenger#fetch), in any case not something particularly meaningful or essential.

My opinion at the moment is: when this is backported to fp-ts v2 (should happen soon), creating and handling `ReaderTaskEither`s is going to be cheaper and more intuitive (no `class` wrapping), so the Fetch name could just be dropped in favor of `ReaderTaskEither`.

Do you have any suggestions in the meantime?

Regarding:
> So the Fetch concept is not strictly related to fetching some data, but to a more general remote call

I'd say not even to the "remote call" part. This is probably not obvious from these docs, but a query is characterized by one or more of the 3 effects described above. E.g. we use it to access `localStorage`, which is only effectful in the sense of IO (sync) and possibly Either (validation of the contents), but definitely not "remote"

```ts
/* when you call the command you can specify the input value corresponding
to the Cache key that should be invalidated as a second parameter */
updatePreferencesCommand({ color: 'acquamarine' }, { preferencesQuery: 1 });
```

# React

> Avenger also exports some utilities to use with `React`

Reviewer 2: Missing dot or colon

## declareQueries

> `declareQueries` is a `HOC` builder. It lets you define the queries that you want to inject into a component and then creates a simple `HOC` to wrap it:

Reviewer 2: is a `HOC` builder --> I would explain what HOC is.

**is a `HOC` (Higher-Order Component) builder**

Reviewer 2: Only after the word appears the first time

        }
      }
    )
  }
}

export queries(MyComponent)

// ./Components/ItemView.ts

class ItemView extends React.PureComponent<Props, State> {
  goToItems: () => doUpdateCurrentView({ view: 'items' }).run()

  render() {
    return <BackButton onClick={this.goToItems}>
  }
}

```

## Signatures

> N.B. all the following signatures reference the abstractions in [`fp-ts`](https://github.com/gcanti/fp-ts)

### `query`
```ts
declare function query<A = void, L = unknown, P = unknown>(
  fetch: Fetch<A, L, P>
): (strategy: Strategy<A, L, P>) => CachedQuery<A, L, P>
```

### `Fetch`

```ts
type Fetch<A, L, P> = (input: A) => TaskEither<L, P>;
```

### `StrategyBuilder`

```ts
type StrategyBuilder<A, L, P> = (
  inputSetoid: Setoid<A>,
  cacheValueSetoid: Setoid<CacheValue<L, P>>
) => Strategy<A, L, P>;
```

### `Strategy`

```ts
export class Strategy<A, L, P> {
  constructor(
    readonly inputSetoid: Setoid<A>,
    readonly filter: Function1<CacheValue<L, P>, boolean>,
    readonly cacheValueSetoid: Setoid<CacheValue<L, P>>
  ) {}
}
```

### `CachedQuery`

```ts
interface CachedQuery<A, L, P> {
  type: 'cached';
  inputSetoid: Setoid<A>;
  run: Fetch<A, L, P>;

>   invalidate: Fetch<A, L, P>;

Reviewer 3: Should `invalidate` be explained somehow?

Author: although exposed, `invalidate` uses are very rare and probably do not deserve to be mentioned here

Building user interfaces is a complex task, mainly because of its `IO` intensive nature. Reads (**queries**) and updates (**commands**) toward "external" data sources are ubiquitous and difficult to orchestrate but _orchestration_ is not the only challenge a UI developer faces, _performance_ and _scalability_ are also key aspects of good design.

We believe that an _effective and powerful abstraction to handle caching and synchronization of external data in a declarative way_ is of fundamental importance when designing a solid user interface.

This is what **Avenger** wants to be: an abstraction layer over external data that handles caching and synchronization for you:

!["cached flow"](docs/Avenger.svg)

By separating how we fetch external data and how we update it we are able to state in a very declarative and _natural_ way the correct lifecycle of that data:

```ts
// define a cached query, with strategy "available" (more about this later)

> const user = queryShallow((id: string) => API.fetchUser(id), available);

Reviewer 4: I know this is probably my own example copy-pasted, but I'd put `queryStrict` here, since the input is a plain primitive value (`string`)

// define a command that invalidates the previous query
const updateUsername = command((patch: Partial<User>) => API.updateUser(patch), { user })

// declare it for usage in a React component
const queries = declareQueries({ user });
const DisplayUsername = queries(props => (
  <div>
    {props.queries.fold(
      () => 'loading...',
      () => 'error while retrieving user',
      queries => (
        <UserNameForm
          value={queries.user.username}
          onSubmit={updateUsername}
        />
      )
    )}
  </div>
));

// render the component

> <Username queries={{ user: '42' }} />;

Reviewer 4: name not in sync with the one used above

```

# Avenger
At the very heart of Avenger's DSL there are two constructors: **query** and **command**.

## queries
The [**`query`**](#query) function allows you to query your data source and get an object of type [**`CachedQuery`**](#CachedQuery) in return.
It accepts two parameters: the first is a function with a [**`Fetch`**](#Fetch) signature that is used to retrieve data from your data source; the second is an object with the[**`Strategy`**](#Strategy) signature that will be used to decide if the data stored by **Avenger** is still relevant or needs to be refetched.

Although important, `query` is a pretty low-level API and **Avenger** offers some convenient utils with a [**`StrategyBuilder`**](#StrategyBuilder) signature that you should prefer over it (unless you have very specific needs):

> - **refetch:** runs the fetch function every time the data is requested.

Reviewer 4: I would add: "(unless there's an ongoing pending request, which is always reused)"

- **expire:** when the data is requested, the fetch function is run only if data in the `Cache` is older than the expiration defined, otherwise the cached value is used.
- **available:** when the data is requested, if a cached value is available it is always returned, otherwise the fetch function is run and the result stored in the `Cache` accordingly.

> All these utils ask you to specify custom [**`Setoid`**](https://github.com/gcanti/fp-ts/blob/master/docs/modules/Setoid.ts.md) instances that will be used to check if an input is already present in one of the `Chache`'s keys (if the check is successful `Avenger` will try to use that value, otherwise it will resort to the `Fetch` function) but you can (and should) use them together with one of the built-in implementations that automatically take care of it:

Reviewer 4: typo in `Chache`

- **queryShallow:** will use a `Setoid` instance that performs a shallow equality check to compare inputs.
- **queryStrict:** will use a `Setoid` instance that performs a strict equality check to compare inputs.
- **queryJSON:** will use a `Setoid` instance that performs a strict equality check after transforming the data via JSON stringification to compare inputs.

Some examples will help clarify:
```ts
/*
  this implementation will always re-run the `Fetch` function
  even if valid cached data is already present
  and use shallow equality to compare input
*/
const myQuery = queryShallow(fetchFunction, refetch);

/*
  this implementation will never run the `Fetch` function
  unless no valid data is present in the Cache
  and use strict equality to compare input
*/
const myQuery = queryStrict(fetchFunction, available);

/*
  this implementation will run the `Fetch` function only if no valid data is present in the Cache
  or t > 10000 ms passed till the last time data was fetched

>   and use strict equality to compare input

Reviewer 4: "JSON equality" actually

*/
const myQuery = queryJSON(fetchFunction, expire(10000));
```

> Each time the `Fetch` function is run with some `inputs`, those same `inputs` are used as a `key` to store the result obtained:

Reviewer 4: I would use the singular `input` here (it's always a single function param)

}
```

From that moment onwards, when **Avenger** will need to decide if the data in our [**`Cache`**](#Cache) is present and still valid it will:

1. attempt to retrieve data from the [**`Cache`**](#Cache)
2. match the result against the cache strategy defined (for instance if we chose `refetch` the data will always be deemed invalid irrespective of the result).

If a valid result is found it is used without further actions, otherwise the `Fetch` function will be re-run in order to get valid data. The two flows are relatively simple:
##### Valid CacheValue
!["cached flow"](docs/CachedValue.svg)

##### Invalid CacheValue
when you call `run` or `subscribe` on a `query` with a combination of `inputs` that was never used before (or whose last use ended up with a `Failure`), avenger will try to run the `Fetch` function resulting in a more complex flow:
!["cached flow"](docs/UncachedOrErrorValue.svg)

There are two ways to get a query result:

```ts
type Error = "500" | "404";
type User = { userName: String };

declare function getUser(userId: number): TaskEither<Error, User>

const userQuery: CachedQuery<number, Error, User> = query(
  getUser,
  refetch

> )(refetch);

Reviewer 4: there's one invocation too much here (remove `(refetch)`)

## composing queries
You can build bigger queries from smaller ones in two ways:

- by composing them with [**`compose`**](#compose): when you need your queries to be sequentially run with the results of one feeding the other, you can use `compose`.
- by grouping them with [**`product`**](#product): when you don't need to run the queries sequentially but would like to conveniently group them and treat them as if they were one you can use `product`*.

*Internally `product` uses the `Applicative` nature of `QueryResults` to group them using the following hierarchical logic:
  1. If any of the queries returned a `Failure` then the whole composition is a `Failure`.
  2. If any of the queries is `Loading` then the whole composition is `Loading`.
  3. If all the queries ended with a `Success` then the composition is a `Success` with a record of results that mirrors the key/value result of the single queries as value.

Here are a couple of simple examples on how to use them:

```ts

> import { compose } from 'avenger/lib/Query';

Reviewer 4: I would say something like "// each value defined is explicitly annotated for clarity, although the annotations are not strictly required"

type UserPreferences = { color: string };

// note that the two ends of the composed functions must have compatible types
declare function getUser(userId: number): TaskEither<Error, User>;
declare function getUserPreferences(
  user: User
): TaskEither<Error, UserPreferences>;

> const userQuery: CachedQuery<number, Error, User> = queryShallow(

Reviewer 4: I'd use `queryStrict` here, same reason as above

  getUser,
  refetch
);

const preferencesQuery: CachedQuery<
  User,
  Error,
  UserPreferences
> = queryShallow(getUserPreferences, refetch);

// this is a query composition
const composition: Composition<number, Error, UserPreferences> = compose(
  userQuery,
  preferencesQuery
);

// this is a query product

> const group: Composition<number, Error, UserPreferences> = product({ myQuery, myQuery2 });

Reviewer 4: type should be `Product<>`
